### ISSUE TITLE
[DB2] : DB2 LUW 10.5 XA (2 phase commit) Errors

### ISSUE THREAD

**Issue by PhilCANDIDO** [atomikos/transactions-essentials#24]
# DB2 Error 2 Phase commits with DB2 LUW 10.5.

## Architecture
We are in microservice architecture, with servers FRONTEND and BACKEND communicating via an messaging application bus (ActiveMQ).

The applications are developed in JAVA, and are hosted by TOMCAT servers.
 
The applications on the BACKEND services, uses the JTA ATOMIKOS with function "2 PHASE COMMIT" to update the data in 2 DB2 databases. Databases use the "HADR" function with a Stanby mode server to secure the data.

## Problem Description
Randomly, you have an error in ATOMIKOS log :
`163012.048//WARNING//com.atomikos.datasource.xa.XAResourceTransaction//XA resource 'BPROD': commit for XID '746D313530373338363631313537393134383932:746D3739313531' raised -4: the supplied XID is invalid for this XA resource//logWarnin`

In sametime we have the error in DB2 diag log file :
```
PID   : 9764         TID : 13344     PROC : db2syscs.exe

INSTANCE: DB2         NODE : 000      DB  : BPROD

APPHDL : 0-23440       APPID: 172.23.3.63.62262.171031194407

AUTHID : BPROD_IDBO_680    HOSTNAME: SDB-DB2-P05

EDUID  : 13344        EDUNAME: db2agent (BPROD) 0

FUNCTION: DB2 UDB, data protection services, sqlpxTEntrySwitchIn, probe:300

MESSAGE : ZRC=0x80100024=-2146435036=SQLP_NOTA "Transaction was not found"

     DIA8036C XA error with request type of "". Transaction was not found.
```

The transaction is not rollback, and no error raised on the application. **So we have an inconsistency between the bases.**

We have open a ticket to IBM DB2 support. After few weeks of investigation, the IBM support answer with the following resultat :
___

1. _In trace jcctraceBPROD.log_xads159, the Xid is tm151179346638104882tm14542, it started a global transaction and trying to commit. The prepare/commit started from Time:2017-11-27-15:37:46.490 as below :_

```
jcc][Time:2017-11-27-15:37:46.396][Thread:messageListenerContainer-1][ResultSet@2c1074a3]closeX (tm151179346638104882tm14542, com.ibm.db2.jcc.t4.T
jcc][Time:2017-11-27-15:37:46.396][Thread:messageListenerContainer-1][ResultSet@2c1074a3]closeX (tm151179346638104882tm14542, com.ibm.db2.jcc.t4.T
jcc][Time:2017-11-27-15:37:46.459][Thread:messageListenerContainer-1][ResultSet@36ba050c]closeX (tm151179346638104882tm14542, com.ibm.db2.jcc.t4.T
jcc][Time:2017-11-27-15:37:46.459][Thread:messageListenerContainer-1][ResultSet@36ba050c]closeX (tm151179346638104882tm14542, com.ibm.db2.jcc.t4.T
jcc][Time:2017-11-27-15:37:46.474][Thread:messageListenerContainer-1][ResultSet@7ec2060b]closeX (tm151179346638104882tm14542, com.ibm.db2.jcc.t4.T
jcc][Time:2017-11-27-15:37:46.474][Thread:messageListenerContainer-1][ResultSet@7ec2060b]closeX (tm151179346638104882tm14542, com.ibm.db2.jcc.t4.T
jcc][Time:2017-11-27-15:37:46.490][Thread:messageListenerContainer-1][T4XAResource@266e0f94]end (tm151179346638104882tm14542, 67108864, 3) called
jcc][Time:2017-11-27-15:37:46.490][Thread:messageListenerContainer-1][T4XAResource@266e0f94]prepare (tm151179346638104882tm14542, 12) called
jcc][Time:2017-11-27-15:37:46.490][Thread:messageListenerContainer-1][T4XAResource@266e0f94]lockForprepare (tm151179346638104882tm14542, 0) called
jcc][Time:2017-11-27-15:37:46.490][Thread:messageListenerContainer-1][T4XAResource@266e0f94]lockForprepare (tm151179346638104882tm14542, 1) called
jcc][Time:2017-11-27-15:37:46.490][Thread:messageListenerContainer-1][T4XAResource@266e0f94]prepareR (tm151179346638104882tm14542, 12) called
jcc][Time:2017-11-27-15:37:46.537][Thread:messageListenerContainer-1][T4XAResource@266e0f94]commit (tm151179346638104882tm14542, false, 7) called
jcc][Time:2017-11-27-15:37:46.537][Thread:messageListenerContainer-1][T4XAResource@266e0f94]lockForcommit (tm151179346638104882tm14542, 0) called
jcc][Time:2017-11-27-15:37:46.537][Thread:messageListenerContainer-1][T4XAResource@266e0f94]lockForcommit (tm151179346638104882tm14542, 1) called
jcc][Time:2017-11-27-15:37:46.537][Thread:messageListenerContainer-1][T4XAResource@266e0f94]commitR (tm151179346638104882tm14542, false, 7) called
```
2. _But at timestamp Time:2017-11-27-15:37:46.506 in trace file jcctraceBPROD.log_xads_1. The TM had called tollback on same Xid tm151179346638104882tm14542_
```
jcc][Time:2017-11-27-15:37:46.506][Thread:Atomikos:1][T4XAResource@1f6747d8]rollback (tm151179346638104882tm14542, 0) called
[jcc][Time:2017-11-27-15:37:46.506][Thread:Atomikos:1][T4XAResource@1f6747d8]lockForrollback (tm151179346638104882tm14542, 0) cal
[jcc][Time:2017-11-27-15:37:46.506][Thread:Atomikos:1][T4XAResource@1f6747d8]lockForrollback (tm151179346638104882tm14542, 1) cal
[jcc][Time:2017-11-27-15:37:46.506][Thread:Atomikos:1][T4XAResource@1f6747d8]rollbackR (tm151179346638104882tm14542, true) called
```

_and its successful, now as per the JDBC XA specifications, the TM invalidated the Xid tm151179346638104882tm14542_

_From the perspective of the resource manager, the pointer xid is valid only for the duration of the call to the xa_rollback() function; that is, once the function ends, the transaction manager is permitted to invalidate where xid points. Resource managers are encouraged to use private copies of *xid after the function ends._

_Due to this invalidation of Xid (Xid gets cleaned on server) , while commiting at 1st point, there is no valid Xid available_

```
DDM Format: d0 RECV(AR) OBJDSS - Object Data Stream Structure
  LL: 18  CORR: 0001  CHAINED: n  CONT ON ERR: n  SAME CORR FOR NEXT DSS: n  

    NM: SYNCCRD - Sync Point Control Reply
      LL: 12    CP: 1248   
      NM: XARETVAL - XA return Value
        LL: 8    CP: 1904   
        XID does not exist (-4)
```
_And thats why driver(Jcc) throws the expected XAException_
```
[jcc][Time:2017-11-27-15:37:46.537][Thread:messageListenerContainer-1][T4XAResource@266e0f94]xaTransactionFinished (0) called
[jcc][Connection@2399077a] DB2 LUWID: 172.23.131.65.61515.171127143319.0E9F
[jcc] BEGIN TRACE_DIAGNOSTICS
[jcc][Thread:messageListenerContainer-1][XAException@40c34f02] javax.transaction.xa.XAException
[jcc][Thread:messageListenerContainer-1][XAException@40c34f02] Message = [jcc][t4][2041][12326][4.23.1044] Erreur lors de l'exꤵtion de XAResource.commit().  Le serveur a renvoy願AER_NOTA. ERRORCODE=-4203, SQLSTATE=null
[jcc][Thread:messageListenerContainer-1][XAException@40c34f02] Stack trace follows
com.ibm.db2.jcc.am.XaException: [jcc][t4][2041][12326][4.23.1044] Erreur lors de l'exꤵtion de XAResource.commit().  Le serveur a renvoy願AER_NOTA. ERRORCODE=-4203, SQLSTATE=null
```

_So this seems the application usage issue ( if you could know why the rollback happened ?) or it needs to be handled at server TM (Transaction Manager) level._
__From Db2 driver (Jcc) point of view, its working as designed.__

_This issue is similar to PMR : 35998,000,738 (about RollbackException after XAER NOTA(-4)) and the conclusion was same here as well._
___

Has anyone ever had a similar problem? Could you help us to resolve the problem?

**Comment by GuyPardon** [comment]
This is a know issue, we have a fix waiting in the 4.0.6 due soon. If it's urgent, please consider purchasing support (in our support branch this was fixed a long time ago).

**Comment by GuyPardon** [comment]
Out of curiosity: did it work for you with 4.0.6?
